fast3d-wgpu-renderer/src/wgpu_device.rs

```diff
@@ -192,20 +195,26 @@ impl<'a> WgpuGraphicsDevice<'a> {
         let mut combine_uniform_buffer_offset = 0;
         let mut frame_uniform_buffer_offset = 0;
 
-        // create shaders in parallel
-        output.draw_calls.par_iter().for_each(|draw_call| {
-            let shared_map = Arc::clone(&self.shader_cache);
-            Self::prepare_shader(device, draw_call.shader_id, draw_call.shader_config, &shared_map);
-        });
+        // create shaders in parallel, but first, reduce to unique shader id's
+        output
+            .draw_calls
+            .par_iter()
+            .map(|draw_call| (draw_call.shader_id, draw_call.shader_config))
```

Comment:
This first `par_iter` might be inefficient and introduce more overhead than it saves in runtime cost.